zkevm-circuits/src/circuit_tools/cell_manager.rs

```diff
@@ -162,12 +318,21 @@ impl<F: Field> CellManager<F> {
                 best_height = column.height;
             }
         }
-        if best_height >= self.height_limit {
-            best_index = None;
+        // Replace a CellType_::Storage by CellType_::StoragePermutation if the later has
+        // better height
```

Comment:
Agreed. This is from evm-circuit as well, they probably do it because there's not much copy constraints. I feel like optimization like this can be done afterwards, at least for a general purpose circuit-tool it shouldn't make such assumption.